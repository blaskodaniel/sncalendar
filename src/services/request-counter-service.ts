import { Injectable, Injector } from '@furystack/inject'
import { IDisposable } from 'monaco-editor'
import { ScopedLogger } from '@furystack/logging'

interface RequestCounterRecord {
  all: number
  GET: number
  POST: number
  sites: {
    [key: string]: {
      all: number
      GET: number
      POST: number
    }
  }
}

@Injectable({ lifetime: 'singleton' })
export class RequestCounterService implements IDisposable {
  private readonly logger: ScopedLogger
  private readonly writeInterval: NodeJS.Timeout
  public dispose() {
    clearInterval(this.writeInterval)
    this.write()
  }

  private readonly defaultState: RequestCounterRecord = {
    all: 0,
    GET: 0,
    POST: 0,
    sites: {},
  }

  private hasChanged: boolean = false

  private currentState: RequestCounterRecord = { ...this.defaultState }

  public reset() {
    this.currentState = { ...this.defaultState }
    this.hasChanged = true
    this.logger.information({ message: 'Request counter reseted.' })
  }

  private getStorageKey() {
    // SN-RequestCounterService-2019-06-17
    return `SN-${this.constructor.name}-${new Date().toISOString().split('T')[0]}`
  }

  public countRequest(site: string, type: 'GET' | 'POST') {
    this.currentState.all += 1
    this.currentState[type] += 1

    if (!this.currentState.sites[site]) {
      this.currentState.sites[site] = {
        all: 0,
        GET: 0,
        POST: 0,
      }
    }

    this.currentState.sites[site].all += 1
    this.currentState.sites[site][type] += 1
    this.hasChanged = true
  }

  public resetToday() {
    localStorage.removeItem(this.getStorageKey())
  }

  private read() {
    try {
      const stored = localStorage.getItem(this.getStorageKey())
      if (stored) {
        this.currentState = JSON.parse(stored)
      }
    } catch (e) {
      // ignore
      this.logger.warning({ message: 'Error reading / parsing request count from local storage' })
    }
  }

  private write() {
    if (this.hasChanged) {
      localStorage.setItem(this.getStorageKey(), JSON.stringify(this.currentState))
      this.hasChanged = false
      this.logger.verbose({ message: `Request count changes persisted. Total: ${this.currentState.all}` })
    }
  }

  constructor(injector: Injector) {
    this.logger = injector.logger.withScope(this.constructor.name)
    this.read()
    this.writeInterval = setInterval(() => {
      this.write()
    }, 5000)
  }
}
