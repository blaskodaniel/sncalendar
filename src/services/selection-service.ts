import { Injectable, Injector } from '@furystack/inject'
import { ScopedLogger } from '@furystack/logging'
import { Disposable, ObservableValue } from '@sensenet/client-utils'
import { GenericContent } from '@sensenet/default-content-types'

/**
 * A context service to get/set the active item and the selection
 */
@Injectable({ lifetime: 'singleton' })
export class SelectionService implements Disposable {
  public activeContent = new ObservableValue<GenericContent | undefined>(undefined)
  public selection = new ObservableValue<GenericContent[]>([])

  private logger: ScopedLogger

  public async dispose() {
    this.activeContent.dispose()
    this.selection.dispose()
  }

  constructor(injector: Injector) {
    this.logger = injector.logger.withScope('SelectionService')

    this.activeContent.subscribe(ac =>
      this.logger.verbose({
        message: ac ? `Active content changed to ${ac.DisplayName || ac.Name}` : `Active content set to None`,
        data: {
          relatedContent: ac,
        },
      }),
    )
    this.selection.subscribe(sel => {
      this.logger.verbose({
        message: sel.length
          ? `Selection changed to: ${sel.map(s => s.DisplayName || s.Name).join(', ')}`
          : `Selection cleared`,
      })
    })
  }
}
