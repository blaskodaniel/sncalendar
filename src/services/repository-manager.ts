import { Injectable, Injector } from '@furystack/inject'
import { FormsAuthenticationService, LoginState, Repository } from '@sensenet/client-core'
import { RepositoryConfiguration } from '@sensenet/client-core/dist/Repository/RepositoryConfiguration'
import { EventHub } from '@sensenet/repository-events'
import { RequestCounterService } from './request-counter-service'

@Injectable({ lifetime: 'singleton' })
export class RepositoryManager {
  private repos: Map<string, Repository> = new Map()
  private eventHubs: Map<string, EventHub> = new Map()
  public getEventHub(repositoryUrl: string) {
    const existing = this.eventHubs.get(repositoryUrl)
    if (existing) {
      return existing
    }
    const instance = new EventHub(this.getRepository(repositoryUrl))
    this.eventHubs.set(repositoryUrl, instance)
    return instance
  }
  public getRepository(repositoryUrl: string, config?: RepositoryConfiguration) {
    const existing = this.repos.get(repositoryUrl)
    if (existing) {
      return existing
    }
    const instance = new Repository(
      {
        ...{
          sessionLifetime: 'expiration',
        },
        requiredSelect: [
          'Id',
          'Path',
          'Name',
          'Type',
          'DisplayName',
          'Icon',
          'IsFolder',
          'ParentId',
          'Version',
          'PageCount' as any,
          'Binary',
          'CreationDate',
          'Avatar',
        ],
        ...config,
        repositoryUrl,
      },
      (input: RequestInfo, init?: RequestInit) => {
        try {
          this.injector
            .getInstance(RequestCounterService)
            .countRequest(new URL(input.toString()).hostname, init && init.method === 'POST' ? 'POST' : 'GET')
        } catch (error) {
          this.injector.getInstance(RequestCounterService).resetToday()
          this.injector.logger.warning({
            scope: 'RepositoryManager',
            message: 'Failed to log the request count :(',
            data: { details: { error } },
          })
        }
        return fetch(input, init)
      },
    )

    FormsAuthenticationService.Setup(instance, {
      select: 'all',
    })
    this.repos.set(repositoryUrl, instance)
    instance.authentication.state.subscribe(s => {
      if (s === LoginState.Authenticated) {
        instance.reloadSchema()
      }
    })
    return instance
  }

  constructor(private readonly injector: Injector) {}
}
