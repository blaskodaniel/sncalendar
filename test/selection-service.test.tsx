import { Injector } from '@furystack/inject'
import { ConsoleLogger } from '@furystack/logging'
import { SelectionService } from '../src/services/selection-service'

describe('SelectionService', () => {
  it('SelectionService snapshot', () => {
    const injector = new Injector()
    injector.options.owner = 'SnApp'
    injector.useLogging(ConsoleLogger)
    const l = new SelectionService(injector)
    expect(l).toMatchSnapshot()
  })
})
