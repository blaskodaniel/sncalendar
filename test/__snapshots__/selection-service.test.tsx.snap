// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectionService SelectionService snapshot 1`] = `
SelectionService {
  "activeContent": ObservableValue {
    "observers": Set {
      ValueObserver {
        "callback": [Function],
        "observable": [Circular],
      },
    },
  },
  "logger": Object {
    "addEntry": [Function],
    "debug": [Function],
    "error": [Function],
    "fatal": [Function],
    "information": [Function],
    "verbose": [Function],
    "warning": [Function],
  },
  "selection": ObservableValue {
    "currentValue": Array [],
    "observers": Set {
      ValueObserver {
        "callback": [Function],
        "observable": [Circular],
      },
    },
  },
}
`;
